gst/nnstreamer/elements/gsttensor_trainer.c

```diff
@@ -471,6 +483,42 @@ gst_tensor_trainer_change_state (GstElement * element,
   return GST_STATE_CHANGE_FAILURE;
 }
 
+
+/**
+ * @brief Check if current epochs is complete, 
+ * tensor_trainer wait for one of epochs to complete before getting the results from the subplugin
+ */
+static gboolean
+gst_tensor_trainer_epochs_is_complete (GstTensorTrainer * trainer)
+{
+  int required_sample;
+
+  g_return_val_if_fail (trainer != NULL, FALSE);
+  g_return_val_if_fail (trainer->fw != NULL, FALSE);
+  g_return_val_if_fail (&trainer->prop != NULL, FALSE);
+
+  required_sample =
+      trainer->prop.num_training_samples + trainer->prop.num_validation_samples;
+  if (trainer->total_push_data_cnt % required_sample != 0)
+    return FALSE;
+
+  g_mutex_lock (&trainer->epoch_complete_lock);
+  if (trainer->is_epoch_complete) {
+    g_mutex_unlock (&trainer->epoch_complete_lock);
+    /* initialize */
+    trainer->is_epoch_complete = FALSE;
+    return TRUE;
```

Comment:
```suggestion
        trainer->is_epoch_complete = FALSE;
        g_mutex_unlock (&trainer->epoch_complete_lock);
    return TRUE;
```

If reading requires locking, writing should require locking, IMO. Is it right?

Reply:
Thank you.